### PR TITLE
ARROW-13226: [Python] Add a general purpose cython trampolining utility

### PR DESCRIPTION
**Pull request by bkietz**

### CODE REVIEW

File: python/pyarrow/tests/test_dataset.py [hunk 1]
```diff
@@ -3397,3 +3404,13 @@ def test_dataset_null_to_dictionary_cast(tempdir, dataset_reader):
...
     table = dataset_reader.to_table(fsds)
     assert table.schema == schema
+
+
+def test_visit_strings_adhoc():
+    import pyarrow._dataset as _ds
+
+    strings = ['a', 'b', 'c']
+    visited = []
+    _ds._visit_strings(strings, visited.append)
```
Comment (pitrou) [inline python/pyarrow/tests/test_dataset.py:3414]: Also test a failing callback?

File: cpp/src/arrow/python/common.h [hunk 2]
```diff
@@ -185,6 +185,62 @@ class ARROW_PYTHON_EXPORT OwnedRefNoGIL : public OwnedRef {
...
+  using Unbound = void(PyObject*, Args...);
+  using Bound = Status(Args...);
+
+  BoundFunction(Unbound* unbound, PyObject* bound_arg)
+      : bound_arg_(bound_arg), unbound_(unbound) {}
+
+  Status Invoke(Args... args) const {
+    unbound_(bound_arg_.obj(), std::forward<Args>(args)...);
+    return CheckPyError();
+  }
```
Comment (pitrou) [inline cpp/src/arrow/python/common.h:204]: If this is meant to be called from arbitrary C++ code, it should probably take the GIL before anything else. Perhaps reuse `SafeCallIntoPython`?

File: cpp/src/arrow/python/common.h [hunk 3]
```diff
@@ -185,6 +185,62 @@ class ARROW_PYTHON_EXPORT OwnedRefNoGIL : public OwnedRef {
...
+std::function<OutFn> BindFunction(Return (*unbound)(PyObject*, Args...),
+                                  PyObject* bound_arg) {
+  using Fn = BoundFunction<Return(PyObject*, Args...)>;
+
+  static_assert(std::is_same<typename Fn::Bound, OutFn>::value,
+                "requested bound function of unsupported type");
+
+  Py_XINCREF(bound_arg);
+  auto bound_fn = std::make_shared<Fn>(unbound, bound_arg);
+  return [bound_fn](Args... args) { return bound_fn->Invoke(args...); };
```
Comment (pitrou) [inline cpp/src/arrow/python/common.h:241]: Should use `std::forward<Args>`?

File: cpp/src/arrow/python/common.h (hunk 2 above)
Comment (bkietz) [reply cpp/src/arrow/python/common.h:204]: SafeCallIntoPython is awkward in the presence of void return, I'll just use PyAcquireGIL

File: python/pyarrow/tests/test_dataset.py (hunk 1 above)
Comment (bkietz) [reply python/pyarrow/tests/test_dataset.py:3414]: will do

File: cpp/src/arrow/python/common.h (hunk 3 above)
Comment (bkietz) [reply cpp/src/arrow/python/common.h:241]: will do

**Review by lidavidm — APPROVED** [review]
Looks reasonable to me.

File: python/pyarrow/tests/test_dataset.py [hunk 4]
```diff
@@ -3095,10 +3095,17 @@ def test_write_dataset_use_threads(tempdir):
         pa.schema([("part", pa.string())]), flavor="hive")
 
     target1 = tempdir / 'partitioned1'
+    paths_written = []
+
+    def writer_pre_finish(writer):
+        pass
+        # paths_written.append(writer.path)
```
[the remark this answers is not in the document]
Comment (bkietz) [reply python/pyarrow/tests/test_dataset.py]: unfortunately there's a segfault I can't track down for this non-example usage

[the remark this answers is not in the document]
Comment (westonpace) [reply python/pyarrow/tests/test_dataset.py]: FYI, I merged this in and modified this part somewhat.  It's working for me at #10628

**Comment by westonpace** [comment]
I've used this as part of #10628 which uses it to hook metadata_collector into the file visitor.  If we want to merge this separately maybe reduce the scope to be just BindFunction?

**Comment by bkietz** [comment]
@westonpace alright I've opened a JIRA just for BindFunction and I'll strip the dataset:: specifics

**Comment by bkietz** [comment]
+1, merging

**Pull request closed without merging**